modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -814,25 +844,106 @@ private CompletableFuture<RowId> takeLocksForUpsert(ByteBuffer searchKey, UUID i
     /**
      * Takes all required locks on a key, before inserting the value.
      *
-     * @param searchKey Key to search.
-     * @param indexId   Index id.
-     * @param txId      Transaction id.
+     * @param tableRow Table row.
+     * @param txId Transaction id.
      * @return Future completes with {@link RowId} or {@code null} if there is no value.
      */
-    private CompletableFuture<RowId> takeLocksForInsert(ByteBuffer searchKey, UUID indexId, UUID txId) {
-        return lockManager.acquire(txId, new LockKey(indexId, searchKey), LockMode.S) // Index S lock
-                .thenCompose(sharedIdxLock -> {
-                    RowId rowId = rowIdByKey(indexId, searchKey);
-
-                    if (rowId == null) {
-                        return lockManager.acquire(txId, new LockKey(indexId, searchKey), LockMode.X) // Index X lock
-                                .thenCompose(exclusiveIdxLock ->
-                                        lockManager.acquire(txId, new LockKey(tableId), LockMode.IX) // IX lock on table
-                                                .thenApply(tblLock -> null));
-                    }
+    private CompletableFuture<?> takeLocksForInsert(BinaryRow tableRow, UUID txId) {
+        return takeLockOnIndexes(tableRow, txId)
+                .thenCompose(ignored -> lockManager.acquire(txId, new LockKey(tableId), LockMode.IX));
+    }
 
-                    return CompletableFuture.completedFuture(rowId);
-                });
+    private CompletableFuture<?> takeLockOnIndexes(BinaryRow tableRow, UUID txId) {
+        List<IndexStorage> indexes = activeIndexes.get();
+
+        if (nullOrEmpty(indexes)) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        CompletableFuture<?>[] locks = new CompletableFuture[indexes.size()];
+        int idx = 0;
+
+        for (IndexStorage index : indexes) {
+            ByteBuffer indexRow = toIndexKey(index, tableRow);
+
+            UUID indexId = indexId(index);
+
+            assert indexRow != null;
+
+            locks[idx++] = lockManager.acquire(txId, new LockKey(indexId, indexRow), LockMode.X);
```

Comment:
Not that simple.Different type of indexes will require different type of locks on different operations on different entries

_different operations_
**Insertion** in sorted non-unique index expect following locks to be taken:
- IX_short(nextKey)
- X_commit(currentKey) if nextKey previously locked in S, X or SIX mode
- IX_commit(currentKey) otherwise

**Select** 
S_commit(currentKey)

_different entries_
More precisely select in non-unique is not always about locking current key, but sometimes is about locking +INF
- If the currentKey is bigger than the upper bound, return NOT_FOUND but take a lock anyway.
- If currentKey matches the upper bound no need to take a lock on the next key.
- If the upper bound is null, take a lock on +INF.
- If nothing is found, take a lock on +INF.

_Different type of indexes_
That part is pretty straight forward, hash index on insert requires different set of locks to be taken in comparison to sorted index, same is about unique/non-unique indexes. 


In order to properly hide given difficulties, I'd favor using Alexey's proposal with wrapping IndexStorages with decorators that'll add index specific lock stuff to put, get and remove methods. That's why, I would expect
`Supplier<List<IndexStorage>> activeIndexes` to return list of such decorator instead of IndexStorage's.